internal/controller/gateway/controller.go

```diff
@@ -67,85 +103,230 @@ func (r *GatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
+					if repoRef.Name == a.GetName() {
+						req = append(req, creconcile.Request{NamespacedName: types.NamespacedName{Namespace: gateway.Namespace, Name: gateway.Name}})
+					}
+				}
+			}
+			return req
+		}),
+	)
+
+	builder.WatchesMetadata(&corev1.Secret{},
+		handler.TypedEnqueueRequestsFromMapFunc(func(ctx context.Context, a client.Object) []creconcile.Request {
+			gatewayList := &securityv1.GatewayList{}
+			listOpts := []client.ListOption{
+				client.InNamespace(a.GetNamespace()),
+			}
+			err := r.List(ctx, gatewayList, listOpts...)
+			if err != nil {
+				if k8serrors.IsNotFound(err) {
+					return []creconcile.Request{}
+				}
+			}
+			req := []creconcile.Request{}
+			for _, gateway := range gatewayList.Items {
+				for _, secretRef := range gateway.Spec.App.ExternalSecrets {
+					if secretRef.Name == a.GetName() {
+						req = append(req, creconcile.Request{NamespacedName: types.NamespacedName{Namespace: gateway.Namespace, Name: gateway.Name}})
+					}
+				}
+				for _, keyRef := range gateway.Spec.App.ExternalKeys {
+					if keyRef.Name == a.GetName() {
+						req = append(req, creconcile.Request{NamespacedName: types.NamespacedName{Namespace: gateway.Namespace, Name: gateway.Name}})
+					}
+				}
+			}
+			return req
+		}),
+	)
+
+	if r.Platform == "openshift" {
+		builder.Owns(&routev1.Route{})
 	}
+	return builder.Complete(r)
+}
 
-	err = reconcile.Services(ctx, params)
+func captureMetrics(ctx context.Context, params reconcile.Params, start time.Time, hasError bool) error {
+
+	gateway := params.Instance
+	operatorNamespace, err := util.GetOperatorNamespace()
 	if err != nil {
-		return ctrl.Result{}, err
+		params.Log.Info("could not determine operator namespace")
+		return err
+	}
+	otelEnabled, err := util.GetOtelEnabled()
+	if err != nil {
+		params.Log.Info("could not determine if OTel is enabled")
+		return err
 	}
 
-	err = reconcile.ServiceAccount(ctx, params)
+	if !otelEnabled {
+		return nil
+	}
+
+	otelMetricPrefix, err := util.GetOtelMetricPrefix()
 	if err != nil {
-		return ctrl.Result{}, err
+		params.Log.Info("could not determine otel metric prefix")
+		return err
 	}
 
-	err = reconcile.Ingress(ctx, params)
+	if otelMetricPrefix == "" {
+		otelMetricPrefix = "layer7_"
+	}
+
+	hostname, err := util.GetHostname()
 	if err != nil {
-		return ctrl.Result{}, err
+		params.Log.Error(err, "failed to retrieve operator hostname")
+		return err
 	}
+	if err != nil {
+		params.Log.Error(err, "failed to retrieve operator namespace")
+		return err
+	}
+
+	meter := otel.Meter("layer7-operator-gateway-controller-metrics")
 
-	err = reconcile.Route(ctx, params)
+	reconcileLatency, err := meter.Float64Histogram(otelMetricPrefix+"operator_gw_reconciler_latency",
+		metric.WithDescription("gateway controller reconcile latency"), metric.WithUnit("ms"))
 	if err != nil {
-		return ctrl.Result{}, err
+		return err
 	}
 
-	err = reconcile.HorizontalPodAutoscaler(ctx, params)
```

Comment:
As its called in look need to have type attribute (operation) for the counters.

Reply:
Updated